lessons/Part2.Rmd

```diff
@@ -0,0 +1,394 @@
+
+```{r}
+## semi_join()
+df1 %>% 
+  semi_join(df2)
+
+## anti_join()
+df1 %>% 
+  semi_join(df2)
+```
+
+3. Set operations treat observations like sets and expect datasets to have the same variables. We are not covering these in this workshop. For more information, consult the `dplyr` manual. 
+
+## Challenge 11: Bulk Recoding Values with a lookup table
+
+Suppose you have a survey that asked three questions (var1,var2, var3) to fifteen subjects.
+
+```{r}
+survey_raw_data <-tibble(
+  ID = LETTERS[1:15],
+  var1 = sample(1:15, replace = F),
+  var2 = sample(16:30, replace = F),
+  var3 = sample(31:45, replace = F)
+)
+glimpse(survey_raw_data)
+```
+
+Survey data is likely to contain errors. Perhaps the enumerators filled in a question incorrectly. Maybe you have a particular value for missing data. As a subject matter expert, you realize that several fixes need to be made globally. You put them into a data frame. 
+
+```{r}
+fixes <- tibble(
+  ID = c("A", "D", "F", "G"),
+  original_var = c("var1", "var2", "var2", "var3"),
+  original_response = as.integer(c(1, 28, 19, 34)),
+  correct_response = as.integer(c(1000, 2800, 1900, 3400))
+)
+glimpse(fixes)
+```
+
+Here we have a problem of updating several variables based on a set of potentially different conditions. Each variable needs to be updated based on the ID, but we do not need to update every variable/ID combination. 
+
+Build a workflow that converts the incorrect answers to correct answers. A solution to this problem makes use of `left_join()`, `mutate()`, `select()`, and both `pivot_longer()` and `pivot_wider()`
+
+```{r}
+
+
+```
+
+## Binding Together Data Frames 
+
+If you know that you have data frames that share columns or rows, you can efficiently turn them into a single data frame with `bind_rows()` or `bind_cols()`. The function that you use depends on the structure of your problem. 
+
+```{r}
+## An Example 
+obs1 <- obs %>% 
+  filter(id == "A")
+
+obs2 <- obs %>% 
+  filter(id == "B")
+
+bind_rows(list(obs1, obs2))
+```
+
+## Storing a single type in multiple tables. 
+
+It is common to find data values about a single observational unit spread out over multiple tables or files. Files are often split by another variable. For example, a set of medical records may be divided by patients. Voting records may be separated by counties or states. Economic data may be separated by economic sector. From a wrangling perspective, as long as there is a consistent format to these records, we can follow a three-step pattern: 
+
+1. Read the files into a list of data frames
+2. For each data frame, add a new column that records the original filename
+3. Combine all the data frames into a single data frame. 
+
+This pattern is so common that it is now built into the `readr` package. 
+
+```{r, eval = F}
+# Get appropriate files. 
+# Here we assume that we have .csv files, but any 
+# support file type will work
+files <- fs::dir_ls(path = "/PATH", glob = "*pattern*csv")
```

Comment:
is there a simple example of this we could use to demonstrate the end result?

Reply:
The challenge is a demonstration of the end result. Do you mean in addition?

Reply:
I see, I missed that. Nevermind, disregard. Thanks for pointing that out.